lib/harmony/harmonytask/harmonytask.go

```diff
@@ -0,0 +1,396 @@
+	//				scream(err)
+	//	 		 	return false
+	//		   }
+	// 		   return true
+	//		})
+	//	  }
+	// }
+	Adder(AddTaskFunc)
+}
+
+type AddTaskFunc func(extraInfo func(TaskID, *harmonydb.Tx) bool)
+
+type TaskEngine struct {
+	ctx            context.Context
+	handlers       []*taskTypeHandler
+	db             *harmonydb.DB
+	workAdderMutex sync.Mutex
+	reg            *resources.Reg
+	grace          context.CancelFunc
+	taskMap        map[string]*taskTypeHandler
+	ownerID        int
+	tryAllWork     chan bool // notify if work completed
+	follows        map[string][]followStruct
+	lastFollowTime time.Time
+	lastCleanup    atomic.Value
+}
+type followStruct struct {
+	f func(TaskID, AddTaskFunc) bool
+	h *taskTypeHandler
+}
+
+type TaskID int
+
+// New creates all the task definitions. Note that TaskEngine
+// knows nothing about the tasks themselves and serves to be a
+// generic container for common work
+func New(
+	db *harmonydb.DB,
+	impls []TaskInterface,
+	hostnameAndPort string) (*TaskEngine, error) {
+
+	reg, err := resources.Register(db, hostnameAndPort)
+	if err != nil {
+		return nil, fmt.Errorf("cannot get resources: %w", err)
+	}
+	ctx, grace := context.WithCancel(context.Background())
+	e := &TaskEngine{
+		ctx:        ctx,
+		grace:      grace,
+		db:         db,
+		reg:        reg,
+		ownerID:    reg.Resources.MachineID, // The current number representing "hostAndPort"
+		taskMap:    make(map[string]*taskTypeHandler, len(impls)),
+		tryAllWork: make(chan bool),
+		follows:    make(map[string][]followStruct),
+	}
+	e.lastCleanup.Store(time.Now())
+	for _, c := range impls {
+		h := taskTypeHandler{
+			TaskInterface:   c,
+			TaskTypeDetails: c.TypeDetails(),
+			TaskEngine:      e,
+		}
+		e.handlers = append(e.handlers, &h)
+		e.taskMap[h.TaskTypeDetails.Name] = &h
+
+		_, err := db.Exec(e.ctx, `INSERT INTO harmony_task_impl (owner_id, name) 
+			VALUES ($1,$2)`, e.ownerID, h.Name)
+		if err != nil {
+			return nil, fmt.Errorf("can't update impl: %w", err)
+		}
+
+		for name, fn := range c.TypeDetails().Follows {
+			e.follows[name] = append(e.follows[name], followStruct{fn, &h})
+
+			// populate harmony_task_follows
+			_, err := db.Exec(e.ctx, `INSERT INTO harmony_task_follows (owner_id, from_task, to_task)
+				VALUES ($1,$2,$3)`, e.ownerID, name, h.Name)
+			if err != nil {
+				return nil, fmt.Errorf("can't update harmony_task_follows: %w", err)
+			}
+		}
+	}
+
+	// resurrect old work
+	{
+		var taskRet []struct {
+			ID   int
+			Name string
+		}
+
+		err := db.Select(e.ctx, &taskRet, `SELECT id, name from harmony_task WHERE owner_id=$1`, e.ownerID)
+		if err != nil {
+			return nil, err
+		}
+		for _, w := range taskRet {
+			// edge-case: if old assignments are not available tasks, unlock them.
+			h := e.taskMap[w.Name]
+			if h == nil {
+				_, err := db.Exec(e.ctx, `UPDATE harmony_task SET owner=NULL WHERE id=$1`, w.ID)
+				if err != nil {
+					log.Error("Cannot remove self from owner field: ", err)
```

Comment:
For errors like this Errorw is better, generally easier to query / track / alert